### ISSUE TITLE
Bring iOS 13 advances/letter spacing in line with iOS < 12.

### ISSUE THREAD

**Issue by GaryQian** [flutter/flutter#43998]
Split off of https://github.com/flutter/flutter/issues/41101

In https://github.com/flutter/flutter/issues/41101, we landed a change (https://github.com/flutter/engine/pull/13321) that uses hb_set_funcs of harfbuzz fonts to change the glyph advance computation to a custom one that is provided by LibTxt and Minikin. This is because the harfbuzz one has difficulty handling fonts in iOS 13.

The new algorithm produces far better results in iOS 13, but we have noticed that on iOS 13, the advances are actually a small amount wider than previous iOS versions. This can lead to unintentional overflow of text.

Here is an example (courtesy of @TheJulianJES):
![image](https://user-images.githubusercontent.com/1887398/68046610-84387880-fc99-11e9-94de-f50a551512da.png)

We should bring all iOS versions back into sync.

**Comment by JagandeepBrar** [comment]
Has there been any update on this? 

With the new v1.12.13+hotfix.5 that was pushed to stable, I am getting a much larger letter spacing on iOS 13.3 developer beta. 

The old stable version looked like the right screenshot, and the new version looks like the left screenshot. A pretty large difference so it’s pretty jarring.

**Comment by TheJulianJES** [comment]
@JagandeepBrar 
As far as I know, the old stable version (v1.9.1+hotfix.6) only had the small letter spacing on iOS 13.
iOS 12 still shows the larger spacing with the old version (v1.9.1+hotfix.6).

In the current stable v1.12.13+hotfix.5 there was a fix to bring back letter spacing in iOS 13.
It's a bit more than it was originally on iOS 12. However, it looks way better than the old small spacing v1.9.1+hotfix.6 had on iOS 13 (and it's very close to the old stable letter spacing on iOS 12).

**Comment by JagandeepBrar** [comment]
@TheJulianJES Yeah, it was just a jarring change when it was less spacing than normal to more than normal. I used `letterSpacing` for the time being to correct it to look as close to iOS 12 as I can for the time being.

**Comment by RabbitKabong** [comment]
In the short term, is there a better way to add a 'global' letterSpacing aside from some nested copyWith to update the body1 TextTheme?

```
ThemeData theme = Theme.of(context).copyWith(
            textTheme: Theme.of(context).textTheme.copyWith(
              body1: Theme.of(context).textTheme.body1.copyWith(letterSpacing: -0.25),
            ),
);
```

**Comment by hishbase** [comment]
The font seems to still be small on iOS devices. However, it looks fine on the simulator. Are you folks not seeing this issue?

**Comment by superakabo** [comment]
> The font seems to still be small on iOS devices. However, it looks fine on the simulator.

For physical devices, you can use this:

double adjustFontSize(BuildContext context, double size) {
  return Platform.isAndroid ? size : (size + (size / MediaQuery.of(context).devicePixelRatio));
}

**Comment by qimingweng** [comment]
Just to add more data points here to compare.

Here (below) is the app running on a physical iPhone XS Max, built with Flutter v1.12.13+hotfix.9

![image](https://user-images.githubusercontent.com/8117770/81234693-44937600-8fc7-11ea-8602-b09b2f837a01.png)

Here (below) is the app running on an iPhone 11 Pro Max Simulator, with Flutter v1.17.0 (stable)

![image](https://user-images.githubusercontent.com/8117770/81234727-5543ec00-8fc7-11ea-9540-75a82b79531d.png)

Finally here is a closer look at each line next to each other (below).

![image](https://user-images.githubusercontent.com/8117770/81234964-e0bd7d00-8fc7-11ea-8c58-0c47d6de5f47.png)

**Comment by JagandeepBrar** [comment]
I have had multiple users give feedback on how the spacing and text looks strange inside of my app after this regression, I'm surprised this doesn't have more priority for a framework designed around "building beautiful user experiences".

Good looking typography is critical and very noticeable in UI.

**Comment by hishbase** [comment]
@JagandeepBrar 💯 

This is a fundamental issue that's been open since November 1. Typography is everything in todays apps. I'm surprised this issue hasn't been resolved in a timely manner. 

Sorry for the harsh comments but we're trying to experiment with Flutter since December and I have not been able to sell this company wide because of this specific issue

**Comment by qimingweng** [comment]
It's been about a month since the last comment so wanted to ask if there was anything new or if there's a separate thread we should be following about this.

**Comment by otopba** [comment]
Same issue

**Comment by TheJulianJES** [comment]
https://github.com/flutter/flutter/issues/60013 might be related.

**Comment by otopba** [comment]
@TahaTesser Any updates?

**Comment by reinhart1010** [comment]
Hi, I decided to research into this and found three things.

First, I think the changes in #41101 should not be the one to blame. In fact, they have rendered the iOS' system fonts in the **correct way**. The major cause behind all of this is the [Material Design's default letter spacing/tracking values](https://m3.material.io/styles/typography/type-scale-tokens) which contradicts [that of Apple](https://developer.apple.com/design/human-interface-guidelines/foundations/typography#specifications). To give some examples:

| Style | Font Size | Material's `letter-spacing` | Apple HIG's `letter-spacing` | Difference |
|---|---|---|---|---|
| Body Large | 16 | 0.5 | -0.31 | **0.81** |
| Body Medium | 14 | 0.25 | -0.15 | **0.4** |
| Body Small | 12 | 0.4 | 0 | **0.4** |

The main reason behind all the expanded text in iOS (and macOS) platforms is simple, the tracking settings for Roboto *et al.* can't be simply applied for San Fransisco.

That's why I decided to add a manual override for these styles. Note that I may not cover all possible text styles in `MaterialApp` for Flutter, like buttons, but this should cover most of the typographical needs for UI elements.

```dart
  /// Adjusts font spacing for iOS and macOS devices
  /// See https://github.com/flutter/flutter/issues/43998
  static ThemeData adjustFontSpacing(ThemeData theme) {
    if (kIsWeb || !shouldUseCupertino()) return theme;
    return theme.copyWith(
      textTheme: theme.textTheme.copyWith(
        bodyLarge: theme.textTheme.bodyLarge?.copyWith(letterSpacing: -0.31),
        bodyMedium: theme.textTheme.bodyMedium?.copyWith(letterSpacing: -0.15),
        bodySmall: theme.textTheme.bodySmall?.copyWith(letterSpacing: 0),
        displayLarge: theme.textTheme.displayLarge?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.29,
        ),
        displayMedium: theme.textTheme.displayMedium?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.35,
        ),
        displaySmall: theme.textTheme.displaySmall?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.37,
        ),
        headlineLarge: theme.textTheme.headlineLarge?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.41,
        ),
        headlineMedium: theme.textTheme.headlineMedium?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.38,
        ),
        headlineSmall: theme.textTheme.headlineSmall?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.07,
        ),
        labelLarge: theme.textTheme.labelLarge?.copyWith(letterSpacing: -0.15),
        labelMedium: theme.textTheme.labelMedium?.copyWith(letterSpacing: 0),
        labelSmall: theme.textTheme.labelSmall?.copyWith(letterSpacing: 0.06),
        titleLarge: theme.textTheme.titleLarge?.copyWith(letterSpacing: -0.26),
        titleMedium: theme.textTheme.titleMedium?.copyWith(letterSpacing: -0.31),
        titleSmall: theme.textTheme.titleSmall?.copyWith(letterSpacing: -0.15),
      ),
    );
  }
```

I've also changed the Display and Heading styles to use Semibold (600) instead of Medium (500) weights.

In conclusion, I'm able to convert from this:

<img width="842" alt="Screenshot 2023-01-10 at 9 37 45 PM" src="https://user-images.githubusercontent.com/17312341/211605541-81890b72-ca98-46b5-ae10-5b3cafb5658d.png">

to this:

<img width="842" alt="Screenshot 2023-01-10 at 9 37 55 PM" src="https://user-images.githubusercontent.com/17312341/211605278-632e852c-f711-4288-b5ed-ba46e46ac437.png">

But there's one more problem: the updated Display and Heading styles are still awkwardly spaced. This is where I learned that **Flutter does not consider using SF Pro Display for text sized at 20px and larger, as recommended and implemented by Apple themselves**. Hence, overriding the font family is also necessary to fix this spacing issue.

```dart
  /// Adjusts font spacing for iOS and macOS devices
  /// See https://github.com/flutter/flutter/issues/43998
  static ThemeData adjustFontSpacing(ThemeData theme) {
    if (kIsWeb || !shouldUseCupertino()) return theme;
    return theme.copyWith(
      textTheme: theme.textTheme.copyWith(
        bodyLarge: theme.textTheme.bodyLarge?.copyWith(letterSpacing: -0.31),
        bodyMedium: theme.textTheme.bodyMedium?.copyWith(letterSpacing: -0.15),
        bodySmall: theme.textTheme.bodySmall?.copyWith(letterSpacing: 0),
        displayLarge: theme.textTheme.displayLarge?.copyWith(
          fontWeight: FontWeight.w600,
          fontFamily: "SF Pro Display",
          letterSpacing: 0.29,
        ),
        displayMedium: theme.textTheme.displayMedium?.copyWith(
          fontWeight: FontWeight.w600,
          fontFamily: "SF Pro Display",
          letterSpacing: 0.35,
        ),
        displaySmall: theme.textTheme.displaySmall?.copyWith(
          fontWeight: FontWeight.w600,
          fontFamily: "SF Pro Display",
          letterSpacing: 0.37,
        ),
        headlineLarge: theme.textTheme.headlineLarge?.copyWith(
          fontWeight: FontWeight.w600,
          fontFamily: "SF Pro Display",
          letterSpacing: 0.41,
        ),
        headlineMedium: theme.textTheme.headlineMedium?.copyWith(
          fontWeight: FontWeight.w600,
          fontFamily: "SF Pro Display",
          letterSpacing: 0.38,
        ),
        headlineSmall: theme.textTheme.headlineSmall?.copyWith(
          fontWeight: FontWeight.w600,
          fontFamily: "SF Pro Display",
          letterSpacing: 0.07,
        ),
        labelLarge: theme.textTheme.labelLarge?.copyWith(letterSpacing: -0.15),
        labelMedium: theme.textTheme.labelMedium?.copyWith(letterSpacing: 0),
        labelSmall: theme.textTheme.labelSmall?.copyWith(letterSpacing: 0.06),
        titleLarge: theme.textTheme.titleLarge?.copyWith(letterSpacing: -0.26),
        titleMedium: theme.textTheme.titleMedium?.copyWith(letterSpacing: -0.31),
        titleSmall: theme.textTheme.titleSmall?.copyWith(letterSpacing: -0.15),
      ),
    );
  }
```

And at last, here's the final result. Quite comparable to the native apps.

<img width="842" alt="Screenshot 2023-01-10 at 9 39 16 PM" src="https://user-images.githubusercontent.com/17312341/211605976-830556e0-45f7-4d52-aa6d-3339a4064ba1.png">

Sure, for a more perfect solution, use CupertinoApp instead for iOS and macOS. But this means developers have to refactor and re-implement their existing styled widgets twice. That's horrible - providing that there are still discussions on whether Flutter will have an official, unified way to use widgets as native as possible, and many existing Flutter apps heavily depend on MaterialApp for custom components.

**Comment by rdesimone** [comment]
IMO, even when using CupertinoApp this issue exists. One must be very careful which widgets to use.

**Comment by reinhart1010** [comment]
Well, I've checked `cupertino/text_theme.dart` and found that they already added correct font families (including text and display) with letter spacing. Can you provide some examples on this?

**Comment by reinhart1010** [comment]
> IMO, even when using CupertinoApp this issue exists. One must be very careful which widgets to use.

Oh, apparently the "SF Pro Display" workaround only works on macOS, and there's another bug report at #82995. I ended up using the "SF Pro Text" instead with adjusted letter spacing

```
  static ThemeData adjustFontSpacing(ThemeData theme) {
    if (!shouldUseCupertino()) return theme;
    return theme.copyWith(
      textTheme: theme.textTheme.copyWith(
        bodyLarge: theme.textTheme.bodyLarge?.copyWith(letterSpacing: -0.31),
        bodyMedium: theme.textTheme.bodyMedium?.copyWith(letterSpacing: -0.15),
        bodySmall: theme.textTheme.bodySmall?.copyWith(letterSpacing: 0),
        displayLarge: theme.textTheme.displayLarge?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.29 - 1.5,
        ),
        displayMedium: theme.textTheme.displayMedium?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.35 - 1.5,
        ),
        displaySmall: theme.textTheme.displaySmall?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.37 - 1.5,
        ),
        headlineLarge: theme.textTheme.headlineLarge?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.41 - 1.5,
        ),
        headlineMedium: theme.textTheme.headlineMedium?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.38 - 1.5,
        ),
        headlineSmall: theme.textTheme.headlineSmall?.copyWith(
          fontWeight: FontWeight.w600,
          letterSpacing: 0.07 - 1.5,
        ),
        labelLarge: theme.textTheme.labelLarge?.copyWith(letterSpacing: -0.15),
        labelMedium: theme.textTheme.labelMedium?.copyWith(letterSpacing: 0),
        labelSmall: theme.textTheme.labelSmall?.copyWith(letterSpacing: 0.06),
        titleLarge: theme.textTheme.titleLarge?.copyWith(letterSpacing: -0.26),
        titleMedium:
            theme.textTheme.titleMedium?.copyWith(letterSpacing: -0.31),
        titleSmall: theme.textTheme.titleSmall?.copyWith(letterSpacing: -0.15),
      ),
    );
  }
```

**Comment by reidbaker** [comment]
Gary is no longer working on flutter issues. Removing him to correctly reflect bug status.

**Comment by MitchellGoodwin** [comment]
I believe this has been fixed by #137275, which changes how we get the native fonts from Apple's APIs. If this is still reproducing, please comment, or open a new issue with reproduction steps.
